[la.solver] enable for std::complex

BarbaraV opened the pull request:
I tried to enable the use of std::complex also for the solvers and first quick tests seem to work. I did not test the implementation for dune-istl, as I haven't yet updated the module and there have been some bugfixes concerning complex in dune-istl lately (around 30/04/2015). 

As the EigenMappedDenseVector only works for double data, I changed the following in the eigen solver:
The old implementation of the class Solver< EigenRowMajorSparseMatrix > is now a template specialization for matrices with double entries. For general EigenRowMajorSparseMatrix, I tried to chekc for inf/nan and symmetry by iterating over non-zero entries (see ll.341 ff. and ll.365 ff.). I'm not sure whether this is a good idea, so please have a look and suggest improvements.

ftalbrecht commented:
Great, thanks for your work!
* regarding istl: as long as it compiles, we can include this, even if it does not yet produce good results. If we know that it does not produce good results with the current istl release, we should add a warning, depending on the version (but this can wait until we have some decent tests for the linear solvers).
* Did you also test the solver in `dune/stuff/la/solver/common.hh`? If it works with complex, could you update it? If not, could you add something like:
```c++
static_assert(!std::is_same< S, std::complex< R > >::value, "Not implemented for std::complex!");
```
* regarding the `EigenRowMajorSparseMatrix` solver: As you already noticed, you duplicated the solver code, which is in general not the best idea. It will make maintenance of the code harder (since if we fix or change one of the solvers, we always have to fix/change the other one accordingly). Since the rest of the solver code is the same for both implementations you could
  * (just for your information) implement a templatized helper struct within the solver, which you use to statically switch only the part of the solver that differs:
```c++
template< class S >
class Solver< EigenRowMajorSparseMatrix< S >, CommunicatorType >
{
  void apply(...)
  {
    // do the general stuff
    ...
    // do the check
    Helper< Stuff::Common::is_complex< S >::value >::check_for_inf_nan(matrix_);
    // do the rest of the general stuff
    ...    
  }

private:
  // general variant (e.g., double)
  template< bool is_complex, bool anything = true >
  struct Helper
  {
    static void check_for_inf_nan(const EigenRowMajorSparseMatrix< S >& matrix)
    {
      // do the old check
      ...
    }
  };

  // specialization for complex
  template< bool is_complex, bool anything >
  struct Helper< true, anything >
  {
    static void check_for_inf_nan(const EigenRowMajorSparseMatrix< S >& matrix)
    {
      // do the new check
      ...
    }
  };
};
```
  * in this case just keep the new variant. My trick using the `EigenMappedDenseVector` was only because it was easiest to write down. The new check you implemented is fine and should work for both complex and non-complex data types.
So I suggest that you completely remove the specialization of the class for `double` and keep the general variant with your newly implemented check.

BarbaraV commented:
Thanks for your suggestions.

I removed the template specialization in Solver< EigenRowMajorSparseMatrix>. If you think that my new implementation is fine, could this check for inf or nan also be a solution to #52 ?

 istl does not compile with std::complex because of wrong return types (real/field type), see https://dune-project.org/flyspray/index.php?do=details&task_id=1622. Therefore, I added a static_assert prohibiting the use of complex.

concerning common: the changes in dune/stuff/la/solver/common.hh seem to be sufficient that the solver works with complex. I tested with
a) 10x10 identity matrix and rhs vector (1+i, 1+i, .....)
b) 10x10 diagonal matrix (1+i)*Id and same rhs vector
In both cases, the expected results are obtained. If you need more information, I can add the corresponding code snippet.

ftalbrecht commented on dune/stuff/la/solver/eigen.hh:
```diff
@@ -337,21 +340,21 @@ public:
-                     << opts);
+      //iterates over the non-zero entries of matrix_.backend() and checks them
+      typedef typename MatrixType::BackendType::InnerIterator InnerIterator;
+      for (int ii = 0; ii < matrix_.backend().outerSize(); ++ii) {
```
Could you please add a private
```c++
typedef typename BackendType::Index EIGEN_size_t;
```
(as in https://github.com/wwu-numerik/dune-stuff/blob/master/dune/stuff/la/container/eigen/sparse.hh#L89), and use `EIGEN_size_t` instead of `int`?

ftalbrecht commented on dune/stuff/la/solver/eigen.hh:
```diff
@@ -361,20 +364,23 @@ public:
+        colmajor_copy -= matrix_.backend().adjoint();
+        //iterates over non-zero entries as above
+        typedef typename ColMajorBackendType::InnerIterator InnerIterator;
+        for (int ii = 0; ii < colmajor_copy.outerSize(); ++ii) {
```
s.a.

ftalbrecht commented on dune/stuff/la/solver/eigen.hh:
```diff
@@ -574,6 +580,8 @@ private:
 
 
+
+
```
Please remove these two lines.

ftalbrecht commented:
Thank you again for your work, I left you some small comments. Regarding istl I think it would be a good idea to guard the `static_assert` in the following way: `#include <dune/common/version.hh>` and guard it with
```c++
#if !DUNE_VERSION_NEWER(DUNE_ISTL, 2, 3)
  static_assert(...);
#endif
```
This should only give a compiler error for istl versions older than 2.4 (the current is 2.3.1, I think). You could then add a todo (as an issue) for us to remember to drop the static_assert as soon as istl releases 2.4. Apart from that everything looks fine and I would merge this soon...

BarbaraV commented:
I changed the eigen-solver according to your suggestions.

I copied your suggestion regarding dune-istl. However, this compiles for me though I have dune-istl version 3.2.1. I'm not quite sure about the macros, but as far as I understood it, does it have to be !DUNE_VERSION_NEWER(DUNE_ISTL, 2, 4) ?

ftalbrecht commented:
According to the docu (http://www.dune-project.org/doc/doxygen/html/version_8hh.html#a3ff6f986ce87d82c20da62e6355cf520) you are right, it should be 2.4.

BarbaraV commented:
thanks, I changes it accordingly and I'll open up a to do in an instant.

As you think that my implementation of the check_for_inf_nan in the eigen solver is fine, shall I change the method valid() for the EigenRowMajorSparseMatrix in the same way? Then this method could be used for all kinds if data (not just double) as well.

ftalbrecht commented:
Thank you, I will merge this soon and get it tested! And yes, please go ahead and fix #52, I completely forgot about that.

ftalbrecht commented:
As always, you can check on the automated builds and tests here, if you like:

http://dune-project.uni-muenster.de/buildbots/dune-stuff/waterfall
